Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift

```diff
@@ -122,150 +141,257 @@ class SentryOnDemandReplay: NSObject, SentryReplayVideoMaker {
+        // Note: In previous implementations this method was wrapped by a sync call to the processing queue.
+        // As this method is already called from the processing queue, we must remove the sync call.
+        let videoFrames = self._frames.filter { $0.time >= beginning && $0.time <= end }
         var frameCount = 0
-        
+
         var videos = [SentryVideoInfo]()
-        
+
         while frameCount < videoFrames.count {
-            let outputFileURL = URL(fileURLWithPath: _outputPath.appending("/\(videoFrames[frameCount].time.timeIntervalSinceReferenceDate).mp4"))
-            if let videoInfo = try renderVideo(with: videoFrames, from: &frameCount, at: outputFileURL) {
-                videos.append(videoInfo)
-            } else {
-                frameCount++
-            }  
+            let frame = videoFrames[frameCount]
+            let outputFileURL = URL(fileURLWithPath: _outputPath)
+                .appendingPathComponent("\(frame.time.timeIntervalSinceReferenceDate)")
+                .appendingPathExtension("mp4")
+
+            let group = DispatchGroup()
+            var currentError: Error?
+
+            group.enter()
+            self.renderVideo(with: videoFrames, from: frameCount, at: outputFileURL) { result in
+                switch result {
+                case .success(let videoResult):
+                    // Set the frame count/offset to the new index that is returned by the completion block.
+                    // This is important to avoid processing the same frame multiple times.
+                    frameCount = videoResult.finalFrameIndex
+                    SentryLog.debug("[Session Replay] Finished rendering video, frame count moved to: \(frameCount)")
+
+                    // Append the video info to the videos array.
+                    // In case no video info is returned, skip the segment.
+                    if let videoInfo = videoResult.info {
+                        videos.append(videoInfo)
+                    }
+                case .failure(let error):
+                    SentryLog.error("[Session Replay] Failed to render video with error: \(error)")
+                    currentError = error
+                }
+                group.leave()
+            }
+
+            // Calling group.wait will block the `processingQueue` until the video rendering completes or a timeout occurs.
+            // It is imporant that the renderVideo completion block signals the group.
+            // The queue used by render video must have a higher priority than the processing queue to reduce thread inversion.
+            // Otherwise, it could lead to queue starvation and a deadlock/timeout.
+            guard group.wait(timeout: .now() + 120) == .success else {
+                SentryLog.error("[Session Replay] Timeout while waiting for video rendering to finish.")
+                throw SentryOnDemandReplayError.errorRenderingVideo
```

Comment:
`m`: If we timeout, we could still return the videos we have and send what we have instead of discarding everything.